File: termpipe_mcp/tools/surgical/readers.py
"""
surgical/readers.py — read-only surgical tools.
Tools: read_lines, find_in_file, read_multiple_files
"""

from pathlib import Path
from typing import Optional
from .helpers import read_file_lines, find_similar_lines


def register_tools(mcp):

    @mcp.tool()
    def read_lines(path: str, start_line: int, end_line: Optional[int] = None) -> str:
        """Read specific line range from a file (0-based)."""
        try:
            lines = read_file_lines(path)
            if end_line is None:
                end_line = start_line + 1
            if start_line < 0 or start_line >= len(lines):
                return f"[Error: Line {start_line} out of range (file has {len(lines)} lines)]"
            output = f"Lines {start_line}-{min(end_line, len(lines))-1} of {path} (total: {len(lines)}):\n\n"
            for i, line in enumerate(lines[start_line:end_line], start_line):
                output += f"{i:4d} | {line}\n"
            return output
        except Exception as e:
            return f"[Error: {e}]"

    @mcp.tool()
    def find_in_file(path: str, pattern: str,
                     max_matches: int = 50, context: int = 0) -> str:
        """Find pattern in file with line numbers and optional context lines."""
        try:
            lines = read_file_lines(path)
            matches = [i for i, l in enumerate(lines)
                       if pattern.lower() in l.lower()][:max_matches]
            if not matches:
                similar = find_similar_lines(lines, pattern)
                if similar:
                    return (f"No exact matches for '{pattern}'\n💡 Similar:\n"
                            + "\n".join(
                                f"  Line {i} ({s:.0%}): {l.strip()[:60]}"
                                for i, l, s in similar))
                return f"No matches for: {pattern}"
            out = f"Found {len(matches)} match(es) for '{pattern}' (file: {len(lines)} lines):\n\n"
            for m in matches:
                if context > 0:
                    s, e = max(0, m - context), min(len(lines), m + context + 1)
                    out += f"--- Line {m} ---\n"
                    for i in range(s, e):
                        out += f"{'→' if i == m else ' '} {i:4d} | {lines[i]}\n"
                    out += "\n"
                else:
                    out += f"Line {m}: {lines[m].strip()[:80]}\n"
            return out
        except Exception as e:
            return f"[Error: {e}]"

    @mcp.tool()
    def read_multiple_files(paths: list[str]) -> str:
        """Read contents of multiple files at once."""
        results = []
        for path in paths:
            p = Path(path).expanduser()
            try:
                if p.suffix.lower() in ['.png', '.jpg', '.jpeg', '.gif', '.webp', '.bmp']:
                    import base64
                    b64 = base64.b64encode(p.read_bytes()).decode()
                    results.append(f"=== {path} ===\n[Image: {p.suffix} ({len(b64)} bytes)]\n")
                else:
                    results.append(f"=== {path} ===\n{p.read_text()}\n")
            except Exception as e:
                results.append(f"=== {path} ===\n[Error: {e}]\n")
        return "\n".join(results)

File: go-mcp/pkg/tools/readers/readers.go
package readers

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func RegisterTools(s *server.MCPServer) {
	// read_lines
	readLinesTool := mcp.NewTool("read_lines",
		mcp.WithDescription("Read specific line range from a file (0-based)."),
		mcp.WithString("path", mcp.Required(), mcp.Description("File path")),
		mcp.WithNumber("start_line", mcp.Required(), mcp.Description("Start line (0-based)")),
		mcp.WithNumber("end_line", mcp.Description("End line (exclusive)")),
	)
	s.AddTool(readLinesTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		path, _ := request.RequireString("path")
		startLine, _ := request.RequireInt("start_line")
		
		endLine := request.GetInt("end_line", startLine+1)

		data, err := os.ReadFile(path)
		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("[Error: %v]", err)), nil
		}

		lines := strings.Split(string(data), "\n")
		if startLine < 0 || startLine >= len(lines) {
			return mcp.NewToolResultText(fmt.Sprintf("[Error: Line %d out of range (file has %d lines)]", startLine, len(lines))), nil
		}

		end := endLine
		if end > len(lines) {
			end = len(lines)
		}

		var output strings.Builder
		output.WriteString(fmt.Sprintf("Lines %d-%d of %s (total: %d):\n\n", startLine, end-1, path, len(lines)))
		for i := startLine; i < end; i++ {
			output.WriteString(fmt.Sprintf("%4d | %s\n", i, lines[i]))
		}

		return mcp.NewToolResultText(output.String()), nil
	})

	// find_in_file
	findInFileTool := mcp.NewTool("find_in_file",
		mcp.WithDescription("Find pattern in file with line numbers and optional context lines."),
		mcp.WithString("path", mcp.Required(), mcp.Description("File path")),
		mcp.WithString("pattern", mcp.Required(), mcp.Description("Search pattern")),
		mcp.WithNumber("max_matches", mcp.Description("Maximum matches")),
		mcp.WithNumber("context", mcp.Description("Context lines")),
	)
	s.AddTool(findInFileTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		path, _ := request.RequireString("path")
		pattern, _ := request.RequireString("pattern")
		pattern = strings.ToLower(pattern)
		
		maxMatches := request.GetInt("max_matches", 50)
		contextLines := request.GetInt("context", 0)

		data, err := os.ReadFile(path)
		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("[Error: %v]", err)), nil
		}

		lines := strings.Split(string(data), "\n")
		var matches []int
		for i, line := range lines {
			if strings.Contains(strings.ToLower(line), pattern) {
				matches = append(matches, i)
				if len(matches) >= maxMatches {
					break
				}
			}
		}

		if len(matches) == 0 {
			return mcp.NewToolResultText(fmt.Sprintf("No matches for: %s", pattern)), nil
		}

		var output strings.Builder
		output.WriteString(fmt.Sprintf("Found %d match(es) for '%s' (file: %d lines):\n\n", len(matches), pattern, len(lines)))
		for _, m := range matches {
			if contextLines > 0 {
				start := m - contextLines
				if start < 0 { start = 0 }
				end := m + contextLines + 1
				if end > len(lines) { end = len(lines) }
				
				output.WriteString(fmt.Sprintf("--- Line %d ---\n", m))
				for i := start; i < end; i++ {
					prefix := " "
					if i == m { prefix = "→" }
					output.WriteString(fmt.Sprintf("%s %4d | %s\n", prefix, i, lines[i]))
				}
				output.WriteString("\n")
			} else {
				line := lines[m]
				if len(line) > 80 { line = line[:80] }
				output.WriteString(fmt.Sprintf("Line %d: %s\n", m, strings.TrimSpace(line)))
			}
		}

		return mcp.NewToolResultText(output.String()), nil
	})

	// read_multiple_files
	readMultipleFilesTool := mcp.NewTool("read_multiple_files",
		mcp.WithDescription("Read contents of multiple files at once."),
		mcp.WithArray("paths", mcp.Required(), mcp.Description("List of file paths")),
	)
	
	s.AddTool(readMultipleFilesTool, func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		paths, err := request.RequireStringSlice("paths")
		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("[Error: %v]", err)), nil
		}

		var output strings.Builder
		for _, path := range paths {
			data, err := os.ReadFile(path)
			if err != nil {
				output.WriteString(fmt.Sprintf("=== %s ===\n[Error: %v]\n", path, err))
				continue
			}

			ext := strings.ToLower(filepath.Ext(path))
			isImage := ext == ".png" || ext == ".jpg" || ext == ".jpeg" || ext == ".gif" || ext == ".webp" || ext == ".bmp"

			if isImage {
				b64 := base64.StdEncoding.EncodeToString(data)
				output.WriteString(fmt.Sprintf("=== %s ===\n[Image: %s (%d bytes)]\n", path, ext, len(b64)))
			} else {
				output.WriteString(fmt.Sprintf("=== %s ===\n%s\n", path, string(data)))
			}
		}

		return mcp.NewToolResultText(output.String()), nil
	})
}
